packages/compass-aggregations/src/modules/pipeline-builder/text-editor.ts

```diff
@@ -0,0 +1,221 @@
+import type { Reducer } from 'redux';
+import type { Document, MongoServerError } from 'mongodb';
+import type { PipelineBuilderThunkAction } from '..';
+import { DEFAULT_MAX_TIME_MS } from '../../constants';
+import type { PreviewOptions } from './pipeline-preview-manager';
+import {
+  DEFAULT_PREVIEW_LIMIT,
+  DEFAULT_SAMPLE_SIZE
+} from './pipeline-preview-manager';
+import { isCancelError } from '../../utils/cancellable-promise';
+import { isAction } from '../../utils/is-action';
+import type { PipelineParserError } from './pipeline-parser/utils';
+import { ActionTypes as PipelineModeActionTypes } from './pipeline-mode';
+import type { PipelineModeToggledAction } from './pipeline-mode';
+import { getStageOperatorsFromPipelineSource } from './builder-helpers';
+import type { PipelineBuilder } from './pipeline-builder';
+
+export const enum EditorActionTypes {
+  EditorPreviewFetch = 'compass-aggregations/pipeline-builder/text-editor/TextEditorPreviewFetch',
+  EditorPreviewFetchSuccess = 'compass-aggregations/pipeline-builder/text-editor/TextEditorPreviewFetchSuccess',
+  EditorPreviewFetchError = 'compass-aggregations/pipeline-builder/text-editor/TextEditorPreviewFetchError',
+  EditorValueChange = 'compass-aggregations/pipeline-builder/text-editor/TextEditorValueChange',
+};
+
+type EditorValueChangeAction = {
+  type: EditorActionTypes.EditorValueChange;
+  pipelineText: string;
+  stageOperators: string[];
+  syntaxErrors: PipelineParserError[];
+};
+
+type EditorPreviewFetchAction = {
+  type: EditorActionTypes.EditorPreviewFetch;
+};
+
+type EditorPreviewFetchSuccessAction = {
+  type: EditorActionTypes.EditorPreviewFetchSuccess;
+  previewDocs: Document[];
+};
+
+type EditorPreviewFetchErrorAction = {
+  type: EditorActionTypes.EditorPreviewFetchError;
+  serverError: MongoServerError;
+};
+
+export type TextEditorState = {
+  pipelineText: string;
+  stageOperators: string[];
+  syntaxErrors: PipelineParserError[];
+  serverError: MongoServerError | null;
+  loading: boolean;
+  previewDocs: Document[] | null;
+};
+
+const INITIAL_STATE: TextEditorState = {
+  pipelineText: '',
+  stageOperators: [],
+  syntaxErrors: [],
+  serverError: null,
+  loading: false,
+  previewDocs: null,
+};
+
+const reducer: Reducer<TextEditorState> = (state = INITIAL_STATE, action) => {
```

Comment:
Seems like actions like loading or creating new empty pipeline (or pipeline from text, but this matters less as we are planning to remove it anyway) are not handled in the reducer, leading to weird behavior